### PR TITLE
Parser 👺👺👺

### PR DESCRIPTION
**Pull request by Furetur**
Bros, this PR is fucked up 🤷‍♂️.

 Pls help me make it more orthodox ☦️

### CODE REVIEW

**Review by achains — CHANGES_REQUESTED** [review]

File: src/flang/parse/parser.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    } else {
+      std::cout << "Unexpected token: " << next_token << "\n";
+      throw parser_exception("Unexpected token");
+    }
+  }
+
+  std::unique_ptr<IdentifierNode> parse_identifier() {
+    auto token = eat(tkIDENTIFIER);
+    auto node = new IdentifierNode(token.value());
+    return std::unique_ptr<IdentifierNode>{node};
```
Comment (achains) [inline src/flang/parse/parser.cpp:44]: std::make_unique

File: src/flang/parse/parser.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+  std::unique_ptr<ProgramNode> parse_program() {
+    std::vector<std::unique_ptr<ElementNode>> elements;
+
+    while (!at_eof()) {
+      auto el = parse_element();
+      elements.push_back(std::move(el));
+    }
+
+    auto node = new ProgramNode(std::move(elements));
```
Comment (achains) [inline src/flang/parse/parser.cpp:25]: std::make_unique

File: src/flang/parse/parser.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  bool at_eof() { return next_index >= tokens.size(); }
+
+  std::vector<Token> &tokens;
+  int next_index;
+};
+
+std::unique_ptr<ProgramNode> parse(std::vector<Token> &tokens) {
+  auto parser = new Parser(tokens);
```
Comment (achains) [inline src/flang/parse/parser.cpp:132]: **Memleak** :) 
Why not `return Parser(tokens).parse_program()`?

**Review by achains — CHANGES_REQUESTED** [review]

File: src/flang/parse/parser.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+#include <flang/tokenize/token.hpp>
+#include <flang/tokenize/token_type.hpp>
+#include <flang/pp.hpp>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+namespace flang {
+
+class Parser {
```
[the remark this answers is not in the document]
Comment (Furetur) [reply src/flang/parse/parser.cpp:13]: Because it's private. Nobody outside of parser.cpp should know that it exists

**Pull request merged**